platform/stm32f4xx/export/platform/irq_manager.hpp

```diff
@@ -13,96 +21,75 @@ using IRQn_t = IRQn_Type;
 class IRQ_manager
 {
 public:
+    using handler_type = std::function< void() >;
+
     // TODO: setup VTOR?
     IRQ_manager() = delete;
     ~IRQ_manager() = delete;
 
-    static void init()
-    {
-        for (auto &h : m_handlers) {
-            h = default_handler;
-        }
-
-        __enable_irq();
-    }
-
-    static int subscribe(IRQn_t IRQn, const std::function< void() > &handler)
-    {
-        // TODO: error check
-
-        __disable_irq();
-
-        // Magic here.
-        // Logical priority of *any* user interrupt that use
-        // FreeRTOS API must not be greather than
-        // configMAX_SYSCALL_INTERRUPT_PRIORITY
-        NVIC_SetPriority(IRQn, CONFIG_MAX_ISR_PRIORITY);
-        m_handlers[IRQn] = handler;
-        __enable_irq();
-        return 0;
-    }
-
-    static int unsubscribe(IRQn_t IRQn)
-    {
-        __disable_irq();
-        m_handlers[IRQn] = default_handler;
-        __enable_irq();
-        return 0;
-    }
-
-    static int mask(IRQn_t IRQn)
-    {
-        // TODO: error check
-        NVIC_DisableIRQ(IRQn);
-        return 0;
-    }
-
-    static int unmask(IRQn_t IRQn)
-    {
-        // TODO: error check
-        NVIC_EnableIRQ(IRQn);
-        return 0;
-    }
-
-    static int clear(IRQn_t IRQn)
-    {
-        // TODO: error check
-        NVIC_ClearPendingIRQ(static_cast< IRQn_t > (IRQn));
-        return 0;
-    }
+    //! Initializes IRQ manager and setups default handlers for every IRQ.
+    static void init();
+
+    //!
+    //! \brief Subscribes to given IRQ.
+    //! \param[in] irqn     Valid IRQ number.
+    //! \param[in] handler  New IRQ handler for given IRQ.
+    //! \retval 0 Success.
+    //!
+    static int subscribe(IRQn_t irqn, const handler_type &handler);
+
+    //!
+    //! \brief Unsubscribes from given IRQ.
+    //! \detail Default handler will be used for given IRQ if this call succeed.
+    //! \param[in] irqn Valid IRQ number.
+    //! \retval 0 Success.
+    //!
+    static int unsubscribe(IRQn_t irqn);
+
+    //!
+    //! \brief Masks or disables given IRQ.
+    //! \param[in] irqn Valid IRQ number.
+    //! \retval 0 Success.
+    //!
+    static int mask(IRQn_t irqn);
+
+    //!
+    //! \brief Unmasks or enables given IRQ.
+    //! \param[in] irqn Valid IRQ number.
+    //! \retval 0 Success.
+    //!
+    static int unmask(IRQn_t irqn);
+
+    //!
+    //! \brief Clears pending interrupt of given IRQ.
+    //! \param[in] irqn Valid IRQ number.
+    //! \retval 0 Success.
+    //!
+    static int clear(IRQn_t irqn);
 
 private:
+    //! Total IRQ count.
+    static constexpr auto irqs = 82;
```

Comment:
we probably need some platform define like PLATFORM_IRQ_COUNT. STM32F40_41xxx and STM32F427_437xx for example has different number of interrupts